controller/api/destination/traffic_split_adaptor.go

```diff
@@ -0,0 +1,63 @@
+package destination
+
+import (
+	"fmt"
+
+	ts "github.com/deislabs/smi-sdk-go/pkg/apis/split/v1alpha1"
+	"github.com/linkerd/linkerd2/controller/api/destination/watcher"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
+)
+
+type trafficSplitAdaptor struct {
+	listener watcher.ProfileUpdateListener
+	id       watcher.ServiceID
+	port     watcher.Port
+	profile  *sp.ServiceProfile
+	split    *ts.TrafficSplit
+}
+
+func newTrafficSplitAdaptor(listener watcher.ProfileUpdateListener, id watcher.ServiceID, port watcher.Port) *trafficSplitAdaptor {
+	return &trafficSplitAdaptor{
+		listener: listener,
+		id:       id,
+		port:     port,
+	}
+}
+
+func (tsa *trafficSplitAdaptor) Update(profile *sp.ServiceProfile) {
+	if tsa.profile != profile {
+		tsa.profile = profile
+		tsa.publish()
+	}
+}
+
+func (tsa *trafficSplitAdaptor) UpdateTrafficSplit(split *ts.TrafficSplit) {
+	if tsa.split != split {
+		tsa.split = split
+		tsa.publish()
+	}
+}
+
+func (tsa *trafficSplitAdaptor) publish() {
+	merged := sp.ServiceProfile{}
+	if tsa.profile != nil {
+		merged = *tsa.profile
+	}
+	if tsa.split != nil {
+		overrides := []*sp.WeightedDst{}
+		for _, backend := range tsa.split.Spec.Backends {
+			dst := &sp.WeightedDst{
+				Authority: fmt.Sprintf("%s.%s.svc.cluster.local:%d", backend.Service, tsa.split.Namespace, tsa.port),
+				Weight:    backend.Weight,
+			}
+			overrides = append(overrides, dst)
+		}
+		merged.Spec.DstOverrides = overrides
```

Comment:
This means any rules manually added into a `ServiceProfile` will get overridden by the corresponding `TrafficSplit` resource. Why not instead of going through service profiles use a separate gRPC endpoint for listening to changes from `TrafficSplit` only? That would also avoid having people install the service's service profile in order to use traffic splitting...

Reply:
I realize now no SP is actually required as `tsa.profile` can be nil, so we're just leveraging the `ServiceProfile` struct. That's cool.